scripts/publish.ts

```diff
@@ -0,0 +1,408 @@
+        "--deferred",
+      ],
+      { dryRun },
+    );
+  }
+  await exec("yarn", ["version", "apply", "--all"], { dryRun });
+
+  // If running in stable mode, both commit and tag the changes. Otherwise, only
+  // tag them.
+  console.log(`🏷️  ${mode === "stable" ? "Committing" : "Tagging"} changes...`);
+  const { stdout: gitDiff } = await exec("git", ["diff", "--name-only"]);
+  if (gitDiff.length === 0 && !dryRun) {
+    throw new CommandError(
+      "There are no changed package versions to commit or tag.",
+    );
+  } else if (mode === "stable") {
+    const commitMessage = `release: ${packageNameToReleaseVersion.size} new package${s}`;
+    await exec("git", ["commit", "--all", "--message", commitMessage], {
+      dryRun,
+    });
+  }
+  for (const [
+    packageName,
+    packageReleaseVersion,
+  ] of packageNameToReleaseVersion) {
+    const gitTagName = `${packageName}@${packageReleaseVersion}`;
+    await exec(
+      "git",
+      ["tag", "--annotate", "--force", "--message", packageName, gitTagName],
+      { dryRun },
+    );
+  }
+
+  // Push the changes to GitHub.
+  console.log("🐱 Pushing to GitHub...");
+  await exec("git", ["push", "--atomic", "--follow-tags"], { dryRun });
+
+  // Publish the packages that are part of the release, both in parallel and in
+  // topological order.
+  console.log("📦 Packing and publishing to npm...");
+  let npmDistributionTag: string;
+  if (mode === "stable") {
+    npmDistributionTag = "latest";
+  } else {
+    npmDistributionTag = isOnDefaultBranch
+      ? maturityLevel
+      : `${maturityLevel}+${sanitizedCurrentBranchName}`;
+  }
+  await exec(
+    "yarn",
+    [
+      ...["workspaces", "foreach", "--all"],
+      "--include",
+      `{${[...packageNameToReleaseVersion.keys()].join(",")}}`,
+      ...["--no-private", "--parallel", "--topological-dev"],
+      ...["npm", "publish"],
+      ...["--tag", npmDistributionTag],
+      ...(args["tolerate-republish"] ? ["--tolerate-republish"] : []),
+    ],
+    { dryRun },
+  );
+
+  // Reset the index and working tree.
+  console.log("🧹 Cleaning...");
+  await exec("git", ["reset", "--hard"], { dryRun });
+
+  // Everything went well.
+  if (dryRun) {
+    console.log("This was a dry run.");
+  } else {
+    const successMessage = `Successfully published ${packageNameToReleaseVersion.size} new package${s}.`;
+    console.log(chalk.green(successMessage));
+  }
+
+  process.off("uncaughtException", handleUncaughtException);
+}
+
+function parsePackageReleaseTypeSpecifier(
+  specifier: string,
+): [name: string, releaseType: string] | undefined {
+  const groups = specifier.match(/^(?<name>\S+):(?<releaseType>\S+)$/)?.groups;
+  return groups?.name !== undefined && groups.releaseType !== undefined
+    ? [groups.name, groups.releaseType]
+    : undefined;
+}
+
+function isReleaseType(value: unknown): value is "major" | "minor" | "patch" {
+  return value === "major" || value === "minor" || value === "patch";
```

Comment:
ℹ️ Btw when doing more than 2 times the same kind of comparaison I like to do that:

```suggestion
  return ["major", "minor", "patch"].includes(value)
```

Reply:
<img width="1666" alt="Screenshot 2023-12-19 at 16 18 45" src="https://github.com/prismicio/slice-machine/assets/8003503/203cc89f-3524-4cda-bc38-afb4cb8876f2">
It doesn't seem to be a good refinement for TypeScript.

Reply:
Arf, forgot the unknown value, good for string yes